url9a01f43c15_poragchowdhury_securitygame_tgz-pJ8-securitygame_GameMasterJ8/src/securitygame/Analyzer.java

```diff
@@ -0,0 +1,298 @@
+        }
+    }
+
+    /**
+     * Outputs Averages
+     */
+    public void printAverages()
+    {
+        System.out.println("Average Points");
+        print(aAverage,attackerNames);
+        print(dAverage,defenderNames);
+    }
+
+    /**
+     * Outputs Medians
+     */
+    public void printMedians()
+    {
+        System.out.println("Medians");
+        print(aMedian, attackerNames);
+        print(dMedian,defenderNames);
+    }
+
+    /**
+     * Outputs Standard Deviations
+     */
+    public void printStandardDev()
+    {
+        System.out.println("Standard Deviations");
+        print(aStd,attackerNames);
+        print(dStd,defenderNames);
+    }
+
+    /**
+     * Outputs Regrets
+     */
+    public void printRegret()
+    {
+        System.out.println("Average Regret");
+        print(aRegret,attackerNames);
+        print(dRegret,defenderNames);
+    }
+
+    /**
+     * Outputs the number of times an agent did the best
+     */
+    public void printBestOf()
+    {
+        System.out.println("Instances Where Agent Was The Best");
+        print(aBestOf,attackerNames);
+        print(dBestOf,defenderNames);
+    }
+
+    /**
+     * Sorts array a and s based on array a
+     * @param a general array of values
+     * @param s names for the elements in a
+     */
+    public void sort(double[] a, String[] s)
+    {
+        for(int i = 0; i < a.length; i++)
+        {
+            for(int j = 0; j < a.length; j++)
+            {
+                if(a[j] < a[i])
+                {
+                    double tempA = a[j];
+                    String tempS = s[j];
+                    a[j] = a[i];
+                    s[j] = s[i];
+                    a[i] = tempA;
+                    s[i] = tempS;
+                }
+            }
+        }
+    }
+
+    /**
+     * Computes the median of array a
+     * @param a general array of values
+     * @return median of a
+     */
+    public double median(int[] a)
+    {
+        int[] m = Arrays.copyOf(a,a.length);
+        Arrays.sort(m);
+        if(m.length%2==0)//even
+            return (m[m.length/2-1]+m[m.length/2])/2.0;
+        else//odd
+            return m[m.length/2];
+    }
+
+    /**
+     * Computes the standard deviation given an array a and an average
+     * @param a general array of values
+     * @param avg average of a
+     * @return standard deviation of a
+     */
+    public double std(int[] a, double avg)
+    {
+        int[] m = Arrays.copyOf(a,a.length);
+        double var = 0;
+        for (int aM : m)
+            var += Math.pow((double) aM - avg, 2);
+        return Math.sqrt(var/m.length);
+    }
+
+    /**
+     * Find the maximum value of a
+     * @param a general array of values
+     * @return maximum value in a
+     */
+    public int maximum(int[] a)
+    {
+        int max = a[0];
+        for(int i = 1; i < a.length; i++)
+            if(max < a[i])
+                max = a[i];
+        return max;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 269:273, 283:287.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.